docs/proposals/container-runtime-interface-v1.md

```diff
@@ -0,0 +1,281 @@
+for the pod, and add/update/remove containers in that environment to meet the
+Pod Spec. To distinguish between the environment and the pod as a whole, we
+will call the pod environment **PodSandbox.**
+
+The container runtimes may interpret the PodSandBox concept differently based
+on how it operates internally. For runtimes relying on hypervisor, sandbox
+represents a virtual machine naturally. For others, it can be Linux namespaces.
+
+In short, a PodSandbox should have the following features.
+
+ * **Isolation**: E.g., Linux namespaces or a full virtual machine, or even
+   support additional security features.
+ * **Compute resource specifications**: A PodSandbox should implement pod-level
+   resource demands and restrictions.
+
+*NOTE: The resource specification does not include externalized costs to
+container setup that are not currently trackable as Pod constraints, e.g.,
+filesystem setup, container image pulling, etc.*
+
+A container in a PodSandbox maps to an application in the Pod Spec. For Linux
+containers, they are expected to share at least network and IPC namespaces,
+with sharing more namespaces discussed in [#1615](https://issues.k8s.io/1615).
+
+
+Below is an example of the proposed interfaces.
+
+```go
+// PodSandboxManager contains basic operations for sandbox.
+type PodSandboxManager interface {
+	Create(config *PodSandboxConfig) (string, error)
+	Delete(id string) (string, error)
+	List(filter PodSandboxFilter) []PodSandboxListItem
+	Status(id string) PodSandboxStatus
+}
+
+// ContainerRuntime contains basic operations for containers.
+type ContainerRuntime interface {
+    Create(config *ContainerConfig, sandboxConfig *PodSandboxConfig, PodSandboxID string) (string, error)
+    Start(id string) error
+    Stop(id string, timeout int) error
+    Remove(id string) error
+    List(filter ContainerFilter) ([]ContainerListItem, error)
+    Status(id string) (ContainerStatus, error)
+	Exec(id string, cmd []string, streamOpts StreamOptions) error
+}
+
+// ImageOperations contains image-related operations.
+type ImageOperations interface {
+    List() ([]Image, error)
+    Pull(image ImageSpec, auth AuthConfig) error
+    Remove(image ImageSpec) error
+    Status(image ImageSpec) (Image, error)
+    Metrics(image ImageSpec) (ImageMetrics, error)
+}
+
+type ContainerMetricsGetter interface {
+    ContainerMetrics(id string) (ContainerMetrics, error)
+}
+
+All functions listed above are expected to be thread-safe.
+```
+
+### Pod/Container Lifecycle
+
+The PodSandbox’s lifecycle is decoupled from the containers, i.e., a sandbox
+is created before any containers, and can exist after all containers in it have
+terminated.
+
+Assume there is a pod with a single container C. To start a pod:
+
+```
+  create sandbox Foo --> create container C --> start container C
+```
+
+To delete a pod:
+
+```
+  stop container C --> remove container C --> delete sandbox Foo
```

Comment:
Ref https://github.com/kubernetes/kubernetes/issues/24677. Also cc @dchen1107 .

Is this meaning that we are turnning back to runtime's own logging API and kubelet does care how runtime managing their container logs?